globals/matchmaking_utils.go

```diff
@@ -140,6 +142,33 @@ func RemoveClientFromAllSessions(client *nex.Client) {
 	}
 }
 
+// CreateSessionByMatchmakeSession creates a gathering from a MatchmakeSession
+func CreateSessionByMatchmakeSession(matchmakeSession *match_making_types.MatchmakeSession, searchMatchmakeSession *match_making_types.MatchmakeSession, hostPID uint32) (*CommonMatchmakeSession, error, uint32) {
+	sessionIndex := GetAvailableGatheringID()
+	// This should in theory be impossible, as there aren't enough PIDs creating sessions to fill the uint32 limit.
+	// If we ever get here, we must be not deleting sessions properly
+	if sessionIndex == 0 {
+		return &CommonMatchmakeSession{}, fmt.Errorf("No gatherings available!"), nex.Errors.RendezVous.LimitExceeded
+	}
+
+	session := CommonMatchmakeSession{
+		SearchMatchmakeSession: searchMatchmakeSession,
+		GameMatchmakeSession:   matchmakeSession,
+	}
+
+	Sessions[sessionIndex] = &session
+	Sessions[sessionIndex].GameMatchmakeSession.Gathering.ID = sessionIndex
+	Sessions[sessionIndex].GameMatchmakeSession.Gathering.OwnerPID = hostPID
+	Sessions[sessionIndex].GameMatchmakeSession.Gathering.HostPID = hostPID
+
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime = nex.NewDateTime(0)
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime.UTC()
+	Sessions[sessionIndex].GameMatchmakeSession.SessionKey = make([]byte, 32)
+    rand.Read(Sessions[sessionIndex].GameMatchmakeSession.SessionKey)
+
+	return Sessions[sessionIndex], nil, 0
```

Comment:
Now that we have `MutexMap` in `nex-go` I wonder if that would be safer here rather than a normal `map`

Reply:
I tried to do this, but wasn't sure how MutexMap is meant to work. Might look into it in a bit, but wanted to get the other changes pushed first :p

Reply:
`MutexMap` is used to create concurrency-safe maps. If we have multiple threads /go routines reading and writing to the map at the same time, Go will panic

`MutexMap` is a generic type which provides a wrapper around a `map` and uses mutex operations to ensure only one thread/go routine accesses the map at a time

Basic usage would be

```go
func main() {
	mutexMap := NewMutexMap[keyType, valueType]()

	mutexMap.Set(key, value)
	value, ok := mutexMap.Get(key)
	mutexMap.Delete(key)
}
```

Reply:
The real map isn't exposed but we have to range through it, so we can't use that atm. Using `Each` is not an option, since on `FindClientSession` we may return early while iterating